dissect/target/plugins/apps/browser/firefox.py

```diff
@@ -265,4 +288,302 @@ def downloads(self) -> Iterator[BrowserDownloadRecord]:
+
+
+def retrieve_master_key(primary_password: bytes, key4_file: TargetPath) -> tuple[bytes, str]:
+    global_salt, password_check = query_global_salt(key4_file)
+    decoded_password_check = core.load(password_check)
+    decrypted_password_check, algorithm = decrypt_master_key(decoded_password_check, primary_password, global_salt)
+
+    if not decrypted_password_check:
+        log.warning("Encountered unknown algorithm %s while decrypting master key.", algorithm)
+        return b"", ""
+
+    expected_password_check = b"password-check\x02\x02"
+    if decrypted_password_check != b"password-check\x02\x02":
+        log.warning(
+            "Master key decryption failed. Either the master key is protected by a "
+            "primary password which you did not supply, or the primary password you provided was incorrect."
+        )
+        log.debug("Expected %s but got %s", expected_password_check, decrypted_password_check)
+        return b"", ""
+
+    master_key, master_key_cka = query_master_key(key4_file)
+    if master_key == b"":
+        log.warning("Password master key is not defined.")
+        return b"", ""
+
+    if master_key_cka != CKA_ID:
+        log.warning("Password master key cka '%s' is not equal to expected value '%s'.", master_key_cka, CKA_ID)
+        return b"", ""
+
+    decoded_master_key = core.load(master_key)
+    decrypted, algorithm = decrypt_master_key(decoded_master_key, primary_password, global_salt)
+    return decrypted[:24], algorithm
+
+
+def decrypt_field(key: bytes, field: tuple[bytes, bytes, bytes]) -> bytes:
+    cka, iv, ciphertext = field
+
+    if cka != CKA_ID:
+        raise ValueError("Expected cka to equal '%s' but got '%s'", CKA_ID, cka)
+
+    return unpad(DES3.new(key, DES3.MODE_CBC, iv).decrypt(ciphertext), 8)
+
+
+def decrypt(
+    username: str, password: str, key4_file: TargetPath, primary_password: str = ""
+) -> tuple[str | None, str | None]:
+    """Decrypt a stored username and password using provided credentials and key4 file.
+
+    Args:
+        username
+        password
+        key4_file
+        primary_password
+
+    Returns: tuple of decoded username and password strings.
+
+    Resources:
+        - https://github.com/lclevy/firepwd
+    """
+    try:
+        username = decode_login_data(username)
+        password = decode_login_data(password)
+
+        primary_password_bytes = primary_password.encode()
+        key, algorithm = retrieve_master_key(primary_password_bytes, key4_file)
+
+        if not key or not algorithm:
+            return None, None
+
+        if algorithm == pbeWithSha1AndTripleDES_CBC or algos.EncryptionAlgorithmId.map(algorithm) == "pbes2":
+            username = decrypt_field(key, username)
+            password = decrypt_field(key, password)
+            return username.decode(), password.decode()
+    except ValueError as e:
+        log.error("Failed to decrypt encrypted password", exc_info=e)
```

Comment:
It's not the prettiest way, but it might be nice to pass in the `target.log` object to these functions. That way the log messages stay bound to the target. There are one or two other places where we do this currently.